moc_openstack_api_reporting.py

```diff
@@ -0,0 +1,586 @@
+#     "vcpus": "1",
+# }
+
+
+def convert_to_ceilometer_event_types(event):
+    map_current_to_old = {
+        "compute.instance.stop": "compute.instance.power_off",
+        "compute.instance.start": "compute.instance.power_on",
+    }
+
+    ceilometer_event_types = {
+        # these messages are reported from compute events
+        # "compute.instance.live-migration": [""],
+        # "compute.instance.detach_volume": [""],
+        # "compute.instance.stop": [""],
+        # "compute.instance.start": [""],
+        # "compute.instance.attach_volume": [""],
+        # "compute.instance.reboot": [""],
+        # "compute.instance.createImage": [""],
+        # "compute.instance.confirmResize": [""],
+        # "compute.instance.migrate": [""],
+        # "compute.instance.resume": [""],
+        # "compute.instance.suspend": [""],
+        # "compute.instance.unshelve": [""],
+        # "compute.instance.shelve": [""],
+        # "compute.instance.unpause": [""],
+        # "compute.instance.pause": [""],
+        # "compute.instance.evacuate": [""],
+        # "compute.instance.attach_interface": [""],
+        # "compute.instance.unlock": [""],
+        # "compute.instance.lock": [""],
+        # "compute.instance.detach_interface": [""],
+        # "compute.instance.rebuild": [""],
+        # "compute.instance.resize": [""],
+        # These events are from the reference file
+        "compute.instance.create": ["start", "end"],
+        "compute.instance.delete": ["start", "end"],  # does this one have a "end"?
+        # "compute.instance.live_migration._post": ["start", "end"],
+        # "compute.instance.live_migration.post.dest": ["start", "end"],
+        # "compute.instance.live_migration.pre": ["start", "end"],
+        # "compute.instance.power_off": ["start", "end"],
+        # "compute.instance.power_on": ["start", "end"],
+        # "compute.instance.shutdown": ["start", "end"],
+    }
+    if event["event_type"] == "compute.instance.create" and event["state"] == "ERROR":
+        event["event_type"] = "compute.instance.create.error"
+        return [event]
+    if event["event_type"] in map_current_to_old:
+        event["event_type"] = map_current_to_old[event["event_type"]]
+    if event["event_type"] in ceilometer_event_types:
+        ret_list = []
+        delta_time = 0
+        time_0 = datetime.datetime.fromisoformat(event["generated"])
+        for e in ceilometer_event_types[event["event_type"]]:
+            new_event = copy.deepcopy(event)
+            new_event["event_type"] = f"{new_event['event_type']}.{e}"
+            new_event["generated"] = (time_0 + datetime.timedelta(seconds=delta_time)).isoformat()
+            ret_list.append(new_event)
+            delta_time += 10
+        return ret_list
+    return []
+
+
+def compile_server_state(server, project_dict, flavor_dict, user_dict):
+    """This one is implied.  We can tell this from the state of the VM"""
+    launched_ts = None
```

Comment:
It doesn't make any sense to set `launched_ts = None` here, since you're setting it unconditionally to another value on the following line (and similar for the `host` and `terminated_ts` variables, below).

Reply:
yes, I didn't do the refactoring using getattr correctly.